### ISSUE TITLE
Anomaly "File "interp/constrintern.ml", line 1817, characters 15-21: Assertion failed."

### ISSUE THREAD

**Issue by afdw** [coq/coq#14708]
#### Description of the problem

1. Download and extract the latest version [Programming Language Foundations](https://softwarefoundations.cis.upenn.edu/plf-current/).
2. Run `make`.
3. Create a file, for example `test123.v`, with the following contents:
```coq
Set Warnings "-notation-overridden,-parsing,-deprecated-hint-without-locality".
From Coq Require Import Bool.Bool.
From PLF Require Import Maps.
From PLF Require Import Smallstep.
From PLF Require Import Stlc.
From PLF Require Import MoreStlc.
Import STLCExtended.

Compute (
  match 123 with
  | <{ \ x : T, t1' }> when true => _
  | _ => 1
  end
).
```
4. Run `coqc -Q . PLF test123.v`.
5. This results in an anomaly:
![image](https://user-images.githubusercontent.com/7559560/126875087-0ca689e7-bf27-40f2-9401-509554793e67.png)

#### Coq Version

```
user@notebook:~$ coqtop -v
The Coq Proof Assistant, version 8.13.2 (May 2021)
compiled on May 28 2021 15:28:51 with OCaml 4.11.1
```

**Comment by JasonGross** [comment]
Let's see if coqbot can minimize this
@coqbot minimize
```
wget https://softwarefoundations.cis.upenn.edu/plf-current/plf.tgz
tar -xzvf plf.tgz
cd plf
make
cat > test123.v <<'EOF'
Set Warnings "-notation-overridden,-parsing,-deprecated-hint-without-locality".
From Coq Require Import Bool.Bool.
From PLF Require Import Maps.
From PLF Require Import Smallstep.
From PLF Require Import Stlc.
From PLF Require Import MoreStlc.
Import STLCExtended.

Compute (
  match 123 with
  | <{ \ x : T, t1' }> when true => _
  | _ => 1
  end
).
EOF
coqc -q -Q . PLF test123.v
```

**Comment by afdw** [comment]
Hm, and I think it can be even simpler than that.
@coqbot minimize
```
wget https://softwarefoundations.cis.upenn.edu/plf-current/plf.tgz
tar -xzvf plf.tgz
cd plf
make
cat > test123.v <<'EOF'
Set Warnings "-notation-overridden,-parsing,-deprecated-hint-without-locality".
From Coq Require Import Bool.Bool.
From PLF Require Import Maps.
From PLF Require Import Smallstep.
From PLF Require Import Stlc.
From PLF Require Import MoreStlc.
Import STLCExtended.

Compute (
  match 123 with
  | <{ unit }> when true => _
  | _ => 1
  end
).
EOF
coqc -q -Q . PLF test123.v
```

**Comment by SkySkimmer** [comment]
Looks like the notation gets around the `such pattern cannot have arguments` error seen eg in `match _ with _ x => _ end`.
For instance 
~~~coq
Notation "'xbar'" := _.
Compute
  match _ with
  | xbar bli => _
  end
.
~~~
gets a similar assert false.

**Comment by herbelin** [comment]
Thanks for the report and the minimization. This was mostly unforeseen and unimplemented cases (since 8.5). Fixed by #14713.
